File: hw/project/src/IBP.R
plot.ibp <- function(Z,show_y_axis=TRUE, show_x_axis=TRUE) {
  N <- nrow(Z)
  K <- ncol(Z)
  image(t(Z), xaxt='n', yaxt='n',fg='grey')
  if (show_x_axis) axis(1,at=(0:(K-1))/(K-1),label=1:K,fg='grey')
  if (show_y_axis) axis(2,at=(0:(N-1))/(N-1),label=N:1,las=1,fg='grey')
}

File: hw/project/src/IBP.jl
module IBP

using Distributions
using RCall
R"source('IBP.R')";
R"library(rcommon)";

export randIBP

function randIBP(alpha::Float64, N::Int, K::Int=20)

  const v = rand(Beta(alpha,1), K)
  const p = cumprod(v,1)
  const Z = [ p[k] > rand() ? 1 : 0 for i in 1:N, k in 1:K ]

  return Z
end

#=
randIBP(1.0, 30, 10)
=#

plot = R"plot.ibp";

@time Zs = [randIBP(5.4,100,30) for i in 1:10000]

stat = [mean(sum(Z,2)) for Z in Zs]
R"plotPost($stat)"

#for i in 1:1000
#  plot(Zs[i],show_y_axis=false, show_x_axis=false)
#end

end
